File: devops-scm-api/src/main/java/com/tencent/devops/scm/api/pojo/webhook/git/CommitCommentHook.java
package com.tencent.devops.scm.api.pojo.webhook.git;

import com.tencent.devops.scm.api.pojo.Commit;
import java.util.Map;
import java.util.Optional;

import com.tencent.devops.scm.api.pojo.Signature;
import com.tencent.devops.scm.api.util.GitUtils;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;

import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_COMMIT_AUTHOR;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_COMMIT_MESSAGE;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_SHA;
import static com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_SHA_SHORT;

@Data
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = false)
@SuperBuilder
public class CommitCommentHook extends AbstractCommentHook {
    public static final String CLASS_TYPE = "commit_comment";
    private Commit commit;

    @Override
    public Map<String, Object> outputs() {
        Map<String, Object> outputs = super.outputs();
        if (commit != null) {
            outputs.put(
                    PIPELINE_GIT_COMMIT_AUTHOR,
                    Optional.ofNullable(commit.getCommitter()).map(Signature::getName).orElse("")
            );
            outputs.put(PIPELINE_GIT_SHA, commit.getSha());
            outputs.put(PIPELINE_GIT_SHA_SHORT, GitUtils.getShortSha(commit.getSha()));
            outputs.put(PIPELINE_GIT_COMMIT_MESSAGE, commit.getMessage());
        }
        if (getExtras() != null) {
            outputs.putAll(getExtras());
        }
        return outputs;
    }
}

File: devops-scm-api/src/main/kotlin/com/tencent/devops/scm/api/pojo/webhook/git/CommitCommentHook.kt
package com.tencent.devops.scm.api.pojo.webhook.git

import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_COMMIT_AUTHOR
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_COMMIT_MESSAGE
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_SHA
import com.tencent.devops.scm.api.constant.WebhookOutputCode.PIPELINE_GIT_SHA_SHORT
import com.tencent.devops.scm.api.enums.EventAction
import com.tencent.devops.scm.api.pojo.Comment
import com.tencent.devops.scm.api.pojo.Commit
import com.tencent.devops.scm.api.pojo.User
import com.tencent.devops.scm.api.pojo.repository.git.GitScmServerRepository
import com.tencent.devops.scm.api.util.GitUtils

data class CommitCommentHook(
    override val action: EventAction,
    override val repo: GitScmServerRepository,
    override val eventType: String,
    override val comment: Comment,
    override val sender: User,
    override var extras: MutableMap<String, Any> = mutableMapOf(),
    val commit: Commit
) : AbstractCommentHook(
    action = action,
    repo = repo,
    eventType = eventType,
    comment = comment,
    sender = sender,
    extras = extras
) {
    companion object {
        const val CLASS_TYPE = "commit_comment"
    }

    override fun outputs(): Map<String, Any> {
        val outputs = super.outputs().toMutableMap()
        outputs[PIPELINE_GIT_COMMIT_AUTHOR] = commit.committer?.name ?: ""
        outputs[PIPELINE_GIT_SHA] = commit.sha
        outputs[PIPELINE_GIT_SHA_SHORT] = GitUtils.getShortSha(commit.sha)
        outputs[PIPELINE_GIT_COMMIT_MESSAGE] = commit.message
        if (extras.isNotEmpty()) {
            outputs.putAll(extras)
        }
        return outputs
    }
}
